### ISSUE TITLE
sshuttle does not work on Fedora 33

### ISSUE THREAD

**Issue by kurtis318** [sshuttle/sshuttle#554]
I have been using sshuttle for about 6 months on Fedora 32 and Debian 10 servers.   Everything was working until I did a scratch install of Fedora 33 on my Dell XPS 15.    The rpm in the Fedora 33 repo is sshuttle-1.0.4-1.fc33.noarch.

From my client I can:
1. Run nslookup ip addresses and fully qualified host names from my client successfully (ip addresses return FQDN and FQDN return ip addresses).
2. I can ssh to a remote machine by ip address.
3. I can open a web page on a remote machine using it's ip address.

I am unable to:
1. Run ssh to any machine by FQDN.
2. Open a web page to any machine by FQDN.

The following command is what I use to establish a connection successfully to a server (just the connected message).
sshuttle --python /home/kurtis318/anaconda3/bin/python --dns -r kurtis318@192.168.0.209 -x 192.168.0.209 0/0

I have tried without specifying --python flag but that does not help.  Anyone got this to work correctly using Fedora 33?  Thanks...

**Comment by svyotov** [comment]
It appears that DNS resolution does not work.

**Comment by skuhl** [comment]
Pull request #552 was merged. Does it fix this issue too?

**Comment by svyotov** [comment]
> Pull request #552 was merged. Does it fix this issue too?

I just tested it and it seems that the issue remains, as Fedora/systemd-resolved is looking somewhere else for the DNS resolution.
When I do `dig private.dns.local` it works before and after the patch, but when accessing though the browser it does not.

When I do :
```
sudo systemd-resolve --interface <interface> --set-dns <DNS-NAMESERVER> --set-domain <private-domain>
```

It does start to work, hope this helps.

**Comment by skuhl** [comment]
Is DNSOverTLS turned on? resolvectl should print this information. If so, does it start to work if you disable DNSOverTLS?

Pull request #552, as mentioned in the commit message, doesn't attempt to address that problem. Although, I think it should be easy enough to fix (we'd have to intercept TCP port 853 connections to the DNS servers).

**Comment by svyotov** [comment]
It's seems like it is not turned on:
```
 $ resolvectl|grep DNSOverTLS
  DNSOverTLS setting: no                  
  DNSOverTLS setting: no  
  DNSOverTLS setting: no                       
  DNSOverTLS setting: no  
  DNSOverTLS setting: no  
  DNSOverTLS setting: no  
```

**Comment by brianmay** [comment]
If it only happens for `*.local` domains, it sounds like you have mdns/avahi enabled (this is the default on modern Linux systems and all OSX systems). Which reserves this domain for its own use. In which case sshuttle can't help, sshuttle currently has no support for mdns.

For more information, see https://www.avahi.org/

**Comment by svyotov** [comment]
> If it only happens for `*.local` domains, it sounds like you have mdns/avahi enabled (this is the default on modern Linux systems and all OSX systems). Which reserves this domain for its own use. In which case sshuttle can't help, sshuttle currently has no support for mdns.
> 
> For more information, see https://www.avahi.org/

This specific issue happens for `marathon.mesos`

**Comment by svyotov** [comment]
> For more information, see https://www.avahi.org/
I will look it up, and when I have more information I will either update here, or create a PR if I find a good solution.
Tnx @brianmay !

**Comment by skuhl** [comment]
It looks like mesos isn't a valid top level domain. It appears that this gets set up by modifying reseolv.conf to be a server that can resolve these domains? https://mesosphere.github.io/mesos-dns/docs/

If you just have one DNS server locally that can resolve those domains, and if you tell sshuttle to send all DNS requests to the remote, then the remote end won't be able to successfully resolve those domains since those DNS requests aren't directed to the DNS server that knows how to do the resolution.

Nonetheless, it sounds like @kurtis318 is describing an issue that is entirely different and @svyotov should perhaps open a separate issue for his issue. For @kurtis318's issue, I did spin up a Fedora 33 server in the cloud. I didn't notice any obvious issues using the latest git master (or the version fedora provides) as long as I remembered to exclude the remote host.

**Comment by brianmay** [comment]
Disregard my comment, that is only applicable for `.local`. And has caught me by surprise in the past when I gave all my local computers names under `.local`, and found nothing worked when friends with OSX computers visited.

If @skuhl is right, then having sshuttle being able to filter DNS requests based on the domain requested might help here (???). I think this is what is being requested in #555. Even if got that wrong, my response I gave there is still applicable here.

**Comment by kurtis318** [comment]
@svyotov I am not sure what to use for the parameters -set-dns and --set-domain in the following command you provided above.  Can you help me here?  Thanks...

sudo systemd-resolve --interface wlp2s0 --set-dns 192.168.88.22 --set-domain yourdomain.local

I am behind a router in my home and my ip address is: 192.168.0.172
My fully qualified hostname is: xps15.roc.mn.charter.com
/etc/resolv.conf is: 
```
# This file is managed by man:systemd-resolved(8). Do not edit.
#
# This is a dynamic resolv.conf file for connecting local clients to the
# internal DNS stub resolver of systemd-resolved. This file lists all
# configured search domains.
#
# Run "resolvectl status" to see details about the uplink DNS servers
# currently in use.
#
# Third party programs should typically not access this file directly, but only
# through the symlink at /etc/resolv.conf. To manage man:resolv.conf(5) in a
# different way, replace this symlink by a static file or a different symlink.
#
# See man:systemd-resolved.service(8) for details about the supported modes of
# operation for /etc/resolv.conf.

nameserver 127.0.0.53
options edns0 trust-ad
search roc.mn.charter.com
```

**Comment by kurtis318** [comment]
I had installed sshuttle from the Fedora 33 repository and it's version is 1.0.4-1.  I uninstalled this version and installed the latest version of sshuttle in this repo and I am able to ssh and open web pages that are only accessible from my remote server.  The version of sshuttle I am running now is 1.0.5.dev65+g7e726bc.   I apologize for not having tried this before opening this issue.  Please close or leave open as you see fit.   

One comment, dns resolution from FQHN to IP address works but not the inverse.  I don't use inverse very often so it's not an issue for me.

**Comment by pimpampeter** [comment]
I just upgraded fedora 32 to 33 and I'm also experiencing the sshuttle issue: the sshuttle --dns  command&parameter combination don't work like before.
The dig and host commands work like expected: the name resolution is succesfully done at the remote end of the ssh tunnel.
curl and webbrowsers (chrome/brave/firefox) don't work like expected (by me) : dns query  isn't send through the tunnel, and results  in dns error (ip address cannot be found).

@kurtis318 : Does the sshuttle version 1.0.5.dev65+g7e726bc solve the issue with --dns , and  where did you find that version?

**Comment by jameslmoser** [comment]
I'm also am having issues with sshuttle on Fedora 33.  Its seems to work but it doesn't work with docker containers anymore, even though it does on Fedora 32.  I have dnsmasq installed locally, and it listens on 0.0.0.0 so that docker containers can use it as well.  With sshuttle running (also with -l 0.0.0.0) with debugging I can see it intercept DNS requests outside of a docker container, but it doesn't when I run a request inside the container.  I removed dnsmasq and tried to use the default systemd-resolved with basically the same results.  I also tried using the newer version from master, and it did not resolve the issue.   I didn't have a lot of time to spend on it,  I was just doing some basic testing in a VM before upgrading my actual workstation, but I would be willing to help debug this if you can tell me any configurations you would like to see or any commands you'd like me to run.

**Comment by jameslmoser** [comment]
Using  docker run --network host  does work, which seems to indicate a problem with docker?? But I'm using the same docker packages for Fedora 32.

**Comment by Obihoernchen** [comment]
Using `1.0.5.dev69+gff70f58` and sshuttle with `--method nft` works totally fine with Fedora 33 :+1:

**Comment by jameslmoser** [comment]
For some reason today it works fine using my regular options...  I don't know what changed.  If it stops working again I will try using `--method nft`.  I think the default is auto, so maybe something changed that was causing it to select something that now and it wasn't before?

**Comment by pimpampeter** [comment]
@Obihoernchen: where did you find this 1.0.5 version of sshuttle ?

**Comment by Obihoernchen** [comment]
Check the readme and install from GitHub directly. Like this:

```
git clone https://github.com/sshuttle/sshuttle.git cd sshuttle sudo ./setup.py install
```

**Comment by pimpampeter** [comment]
Thank you , I'll try that. Since I couldn't find in the source any version info and no news in the changelog, I presumed this was still 1.04.

**Comment by birkenfeld** [comment]
I had the same problem, and traced back to this entry in `/etc/nsswitch.conf`:
```
hosts:      files mdns4_minimal [NOTFOUND=return] resolve [!UNAVAIL=return] myhostname dns
```

The `resolve` entry uses `libnss_resolve.so` for name resolution, with directly queries `systemd-resolved` - this in turn seems to use trigger some resolution method that is not redirected by sshuttle.

When I removed `resolve [!UNAVAIL=return]`, it worked fine with version 1.0.4 which is in Fedora 33.

**Comment by skuhl** [comment]
As far as I know, the fix for issue #535 (pull request #552) should fix the issue described in the previous comment (unless something like DNS over TLS is used).

**Comment by Germano0** [comment]
You may want to test 1.0.5, it solved many internal domain resolve problems I was experiencing.
I created a Copr repository for Fedora 33
https://copr.fedorainfracloud.org/coprs/germano/sshuttle_f33/

**Comment by rderooy** [comment]
FYI, at least for F34 beta the 1.0.5 version is in the repo.

```
$ rpm -qi sshuttle
Name        : sshuttle
Version     : 1.0.5
Release     : 2.fc34
Architecture: noarch
Install Date: Tue 23 Mar 2021 08:11:06 CET
Group       : Unspecified
Size        : 452459
License     : LGPLv2+
Signature   : RSA/SHA256, Fri 29 Jan 2021 04:40:44 CET, Key ID 1161ae6945719a39
Source RPM  : sshuttle-1.0.5-2.fc34.src.rpm
Build Date  : Fri 29 Jan 2021 04:25:25 CET
Build Host  : buildvm-a32-15.iad2.fedoraproject.org
Packager    : Fedora Project
Vendor      : Fedora Project
URL         : https://github.com/sshuttle/sshuttle
Bug URL     : https://bugz.fedoraproject.org/sshuttle
Summary     : Transparent Proxy VPN
Description :
Transparent proxy server that works as a poor man's VPN. Forwards over ssh.
Doesn't require admin. Works with Linux and MacOS. Supports DNS tunneling.
```

**Comment by rderooy** [comment]
One other issue that I encountered with systemd-resolve is it's cache. If you try to resolve a name before starting sshuttle, it will keep failing to resolve it after starting sshuttle. As such before starting sshuttle you should run:
```
sudo systemd-resolve --flush-caches
```

**Comment by rderooy** [comment]
Also after terminating sshuttle you really need to flush the cache. As such I made a batch file for my common use case.

```
$ cat ~/bin/sshuttle.sh 
#!/bin/sh

# catch CTRL-C so we can clean up
trap ctrl_c INT

function ctrl_c() {
	sudo systemd-resolve --flush-caches
	echo "Flushed DNS cache"
}

# set window title to 'sshuttle'
echo -ne "\033]0;sshuttle\007"

# flush dns cache
ctrl_c

sshuttle --dns -r login.example.com:2222 10.0.0.0/8
```

**Comment by skuhl** [comment]
Does the code in pull request #634 fix part of (or all of?) this issue? It flushes the systemd resolver cache on startup and on exit.

**Comment by rderooy** [comment]
I tried it here, and it seems to do the trick!

I tested it by doing:

- nslookup host.domain (failed)
- sshuttle....
- nslookup host.domain (success)
- visit host.domain webpage (success)
- sent ctrl-c to sshuttle
- nslookup host.domain (failed)
